File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *


input = stdin.readline

INF = 1000001

n,k = map(int,input().split())


dist = [inf for _ in range(INF)]
visited = [False for _ in range(INF)]


heap = []

dist[n] = 0
heappush(heap,(0,n))
while heap:
    
    head = heappop(heap)
    currDist = head[0]
    currPos = head[1]
    
    
    if currPos == k:
        break
    
    if visited[currPos]:
        continue
    
    visited[currPos] = True
    
    right = currPos + 1
    left = currPos - 1
    warp = 2 * currPos
    
    if warp < INF and dist[warp] > currDist +1:
        dist[warp] = currDist
        heappush(heap,(dist[warp],warp))
    
    if left >=0 and dist[left] > currDist + 1:
        dist[left] = currDist + 1
        heappush(heap,(dist[left],left))
    
    if right < INF and dist[right] > currDist +1:
        dist[right] = currDist + 1
        heappush(heap,(dist[right],right))
    



print(dist[k])
    

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.swift
struct Edge{
    let start:Int
    let cost:Int
    
  
    public static func <=(lhs:Self,rhs:Self) -> Bool {
        return rhs.cost <= lhs.cost
    }
}


struct Heap<T>{
    
    var nodes:[T] = []
    var comparer:(T,T) -> Bool
    
    var isEmpty:Bool {
        return nodes.isEmpty
    }
    
    var top:T? {
        return nodes.first
    }
    
    var count:Int {
        return nodes.count
    }
    
    init(comparer: @escaping (T,T) -> Bool ){
        self.comparer = comparer
    }
    
    
    mutating func push(_ element:T){
        
        nodes.append(element)
        
        var index:Int = nodes.count-1
        
        
        while index > 0 && !comparer(nodes[index],nodes[(index-1)/2]) {
            
            let parent = (index-1)/2
            nodes.swapAt(index, parent)
            index = parent
            
        
        }
        
    }
    
    mutating func pop() ->T? {
        
        guard !nodes.isEmpty else {
            return nil
        }
        
        if nodes.count == 1 {
            return nodes.popLast()
        }
        
        
        nodes.swapAt(0, nodes.count - 1)
        
        let result = nodes.popLast()
        
        var index:Int = 0
        
        
        while index < nodes.count {
            
            let left:Int = index * 2+1
            let right:Int = left+1
            
            
            if right < nodes.count {
                
                if comparer(nodes[left],nodes[right]) && !comparer(nodes[right],nodes[index]) {
                    
                    nodes.swapAt(right, index)
                    index = right
                    
                }
                
                else if !comparer(nodes[left],nodes[index]){
                    nodes.swapAt(left, index)
                    index = left
                }
                
                else {
                    break
                }
                
            }
            
            else if left < nodes.count {
                if !comparer(nodes[left],nodes[index]){
                    nodes.swapAt(left, index)
                    index = left
                }
                else {
                    break
                }
            }
            else {
                break
            }
            
        }
        
        
        return result
    }
    
    
}

let nk = readLine()!.split{$0 == " "}.compactMap{Int(String($0))!}

let n = nk[0]
let k = nk[1]


var heap = Heap<Edge>(comparer: <=)

heap.push(Edge(start: n, cost: 0))
var ans:Int = 0

let MAX:Int = 100001

var dist:[Int] = [Int](repeating:Int.max,count:MAX)

while !heap.isEmpty {
    
    let top = heap.pop()!
    
    if top.start == k {
        ans = top.cost
        break
    }
    
    let right = top.start+1
    
    
    let left = top.start-1
    
    let warp = top.start * 2
    
    if right < MAX &&  dist[right] > top.cost + 1{
        dist[right] = top.cost + 1
        heap.push(Edge(start: right, cost: dist[right]))
    }
    
    if left >= 0 && dist[left] > top.cost + 1 {
        dist[left] = top.cost + 1
        heap.push(Edge(start: left, cost: dist[left]))
    }
    
    if warp < MAX && dist[warp] > top.cost{
        dist[warp] = top.cost // 순간이동은 그대로
        heap.push(Edge(start: warp, cost: dist[warp]))
    }
    
}


print(ans)
